### PR TITLE
[Backport] Don't load product collection in review observer

### PR DESCRIPTION
**Pull request by Den4ik**
### Original Pull Request
https://github.com/magento/magento2/pull/21200

### Description (*)

Module observers executed after `CatalogBlockProductCollectionBeforeToHtmlObserver` don't have affect on collection. This PR is changing Review observer logic. If fact we can join only 2 reviews_count and rating_summary fields instead of send additional sql query for retrieving this data.

### Fixed Issues (if relevant)

Changed append summary to product collection to mysql join
1. magento/community-features#57: DO NOT do `load()` of product collection in `Review` observer as it causes issues for other observers

### Manual testing scenarios (*)

1. Create new module
2. Add event observer for `catalog_block_product_list_collection`
3. Do some changes to collection in observer

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Review by orlangur — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-5199](https://jira.corp.magento.com/browse/ENGCOM-5199) has been created to process this Pull Request

File: app/code/Magento/Review/Model/ReviewSummary.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Append review summary data to product
+     *
+     * @param AbstractModel $object
+     * @param int $storeId
+     * @param int $entityType
+     */
+    public function appendSummaryDataToObject(AbstractModel $object, int $storeId, int $entityType = 1): void
```
Comment (ysapiga) [inline app/code/Magento/Review/Model/ReviewSummary.php]: @Den4ik please remove return type. Magento 2.2.x support php 7.0 but 'void' return type available only from php 7.1

**Comment by Den4ik** [comment]
@ysapiga Done

**Comment by engcom-Golf** [comment]
@orlangur please approve changes

**Comment by Den4ik** [comment]
@orlangur 
> please approve changes

**Review by sidolov — APPROVED** [review]

File: lib/internal/Magento/Framework/Data/Collection/AbstractDb.php [hunk 2]
```diff
@@ -217,7 +218,7 @@ public function getSize()
     {
         if ($this->_totalRecords === null) {
             $sql = $this->getSelectCountSql();
-            $this->_totalRecords = $this->getConnection()->fetchOne($sql, $this->_bindParams);
+            $this->_totalRecords = $this->_totalRecords ?? $this->getConnection()->fetchOne($sql, $this->_bindParams);
```
Comment (korostii) [inline lib/internal/Magento/Framework/Data/Collection/AbstractDb.php:221]: It seems that when if reach this point, $this->_totalRecords must be strictly null at the time null coalescence check. Or am i missing something?

@orlangur @Den4ik

Comment (p-bystritsky) [reply lib/internal/Magento/Framework/Data/Collection/AbstractDb.php:221]: @korostii no. Value can be set during getSelectCountSql method execution, and Elasticsearch adapter does it at some point.

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**